lib/ansible/modules/cloud/vmware/vmware_guest_disk_facts.py

```diff
@@ -0,0 +1,196 @@
+   folder:
+     description:
+     - Destination folder, absolute or relative path to find an existing guest.
+     - This is required parameter, only if multiple VMs are found with same name.
+     - The folder should include the datacenter. ESX's datacenter is ha-datacenter
+     - 'Examples:'
+     - '   folder: /ha-datacenter/vm'
+     - '   folder: ha-datacenter/vm'
+     - '   folder: /datacenter1/vm'
+     - '   folder: datacenter1/vm'
+     - '   folder: /datacenter1/vm/folder1'
+     - '   folder: datacenter1/vm/folder1'
+     - '   folder: /folder1/datacenter1/vm'
+     - '   folder: folder1/datacenter1/vm'
+     - '   folder: /folder1/datacenter1/vm/folder2'
+     - '   folder: vm/folder2'
+     - '   folder: folder2'
+   datacenter:
+     description:
+     - The datacenter name to which virtual machine belongs to.
+     required: True
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
+- name: Gather disk facts from virtual machine using UUID
+  vmware_guest_disk_facts:
+    hostname: 192.168.1.209
+    username: administrator@vsphere.local
+    password: vmware
+    datacenter: ha-datacenter
+    validate_certs: no
+    uuid: 421e4592-c069-924d-ce20-7e7533fab926
+  delegate_to: localhost
+  register: disk_facts
+
+- name: Gather disk facts from virtual machine using name
+  vmware_guest_disk_facts:
+    hostname: 192.168.1.209
+    username: administrator@vsphere.local
+    password: vmware
+    datacenter: ha-datacenter
+    validate_certs: no
+    name: VM_225
+  delegate_to: localhost
+  register: disk_facts
+'''
+
+RETURN = """
+guest_disk_facts:
+    description: metadata about the virtual machine's disks
+    returned: always
+    type: dict
+    sample: {
+        "0": {
+            "backing_datastore": "datastore2",
+            "backing_disk_mode": "persistent",
+            "backing_eagerlyscrub": false,
+            "backing_filename": "[datastore2] VM_225/VM_225.vmdk",
+            "backing_thinprovisioned": false,
+            "backing_writethrough": false,
+            "capacity_in_bytes": 10485760,
+            "capacity_in_kb": 10240,
+            "controller_key": 1000,
+            "key": 2000,
+            "label": "Hard disk 1",
+            "summary": "10,240 KB",
+            "unit_number": 0
+        },
+    }
+"""
+
+try:
+    from pyVmomi import vim
+except ImportError:
+    pass
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.vmware import PyVmomi, vmware_argument_spec
+
+
+class PyVmomiHelper(PyVmomi):
+    def __init__(self, module):
+        super(PyVmomiHelper, self).__init__(module)
+
+    def gather_disk_facts(self, vm_obj):
+        """
+        Function to gather facts about VM's disks
+        Args:
+            vm_obj: Managed object of virtual machine
+
+        Returns: A list of dict containing disks information
+
+        """
+        disks_facts = dict()
```

Comment:
This should be a list, not a dictionary. The dictionary key is just an incrementing index value, which is what a list is for. Having it as a dictionary just makes it harder to use (having to deal with an index generator, and dictionaries don't guarantee order).

Reply:
Sorry, but I forgot the reason why I implemented dict instead of list for returning facts. If you feel that list is more appropriate than dict, then we can always refactor this code again. PR welcomed.

Reply:
@phemmer , We've discussed this in the weekly Ansible VMware meeting. 
In my opinion, having a dict is useful, IF the key is useful:
* The reason to have facts is to use them in a task later on
* With a list, to find the proper disk, it is harder to find the item you want to match on ( `{{ (disks |  selectattr('label', 'search', 'Hard disk 1') | list | first) }}` for instance)
* With a dict, if the key is useful, it doesn't require that. If you want to match on another criteria, the loop is still a possibility.

Proposal is to have the label as the dict key as it is what is visible in vCenter and is orderable, does that make sense to you? Any other criteria you might find easier? (disk.key is another possibility, but less 'visible' in vCenter)